runtime/runtime_impl_jetty9/src/main/java/com/google/apphosting/runtime/jetty9/JettyHttpHandler.java

```diff
@@ -0,0 +1,290 @@
+
+    // Set the environment as a request attribute, so it can be pulled out and set for async
+    // threads.
+    ApiProxy.Environment currentEnvironment = ApiProxy.getCurrentEnvironment();
+    request.setAttribute(AppEngineConstants.ENVIRONMENT_ATTR, currentEnvironment);
+
+    try {
+      dispatchRequest(target, requestToken, genericRequest, genericResponse);
+      if (!baseRequest.isHandled()) {
+        response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "request not handled");
+      }
+    } catch (
+        @SuppressWarnings("InterruptedExceptionSwallowed")
+        Throwable ex) {
+      // Note we do intentionally swallow InterruptException.
+      // We will report the exception via the rpc. We don't mark this thread as interrupted because
+      // ThreadGroupPool would use that as a signal to remove the thread from the pool; we don't
+      // need that.
+      handleException(ex, requestToken, genericResponse);
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, ex.getMessage());
+    } finally {
+      requestManager.finishRequest(requestToken);
+    }
+    // Do not put this in a final block.  If we propagate an
+    // exception the callback will be invoked automatically.
+    genericResponse.finishWithResponse(context);
+    // We don't want threads used for background requests to go back
+    // in the thread pool, because users may have stashed references
+    // to them or may be expecting them to exit.  Setting the
+    // interrupt bit causes the pool to drop them.
+    if (genericRequest.getRequestType() == RuntimePb.UPRequest.RequestType.BACKGROUND) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  private void dispatchRequest(
+      String target,
+      RequestManager.RequestToken requestToken,
+      JettyRequestAPIData request,
+      JettyResponseAPIData response)
+      throws Throwable {
+    switch (request.getRequestType()) {
+      case SHUTDOWN:
+        logger.atInfo().log("Shutting down requests");
+        requestManager.shutdownRequests(requestToken);
+        request.getBaseRequest().setHandled(true);
+        break;
+      case BACKGROUND:
+        dispatchBackgroundRequest(request, response);
+        request.getBaseRequest().setHandled(true);
+        break;
+      case OTHER:
+        dispatchServletRequest(target, request, response);
+        break;
+      default:
+        throw new IllegalStateException(request.getRequestType().toString());
+    }
+  }
+
+  private void dispatchServletRequest(
+      String target, JettyRequestAPIData request, JettyResponseAPIData response)
+      throws Throwable {
+    Request baseRequest = request.getBaseRequest();
+    HttpServletRequest httpServletRequest = request.getHttpServletRequest();
+    HttpServletResponse httpServletResponse = response.getHttpServletResponse();
+    baseRequest.setAttribute(AppEngineConstants.APP_VERSION_KEY_REQUEST_ATTR, appVersionKey);
```

Comment:
Do we need another attribute? can't this be recovered from the ENVIRONMENT _ATTR?

Reply:
This attribute is already used in the `AppVersionHandlerMap` and I don't think we want to change this for the jetty9 runtime. I think its not really used anymore in the jetty12 version though.